pkg/internal/controller/controller.go

```diff
@@ -203,19 +203,17 @@ func (c *Controller) Start(ctx context.Context) error {
 		// which won't be garbage collected if we hold a reference to it.
 		c.startWatches = nil
 
-		if c.JitterPeriod == 0 {
-			c.JitterPeriod = 1 * time.Second
-		}
-
 		// Launch workers to process resources
 		c.Log.Info("Starting workers", "worker count", c.MaxConcurrentReconciles)
+		wg.Add(c.MaxConcurrentReconciles)
 		for i := 0; i < c.MaxConcurrentReconciles; i++ {
-			go wait.UntilWithContext(ctx, func(ctx context.Context) {
```

Comment:
From what I can see, the `wait.UntilWithContext` was harmless, but not useful because it would start the closure every c.JitterPeriod for as long as the ctx is not cancelled. The closure will however stay in the `for c.processNextWorkItem(ctx)` loop until the workqueue is shutdown and the workqueue is only shutdown when the `ctx` is cancelled.

I think we still have this because before b4a0212360dfc048233a01ee193de7b1d4dd1ab4 it was possible for the closure to end.

With the changes in this PR, using it may lead to a deadlock:
* We add the workers to the `WaitGroup`
* Context gets cancelled before the closure is started
* `wait.UntilWithContext` will not start the closure so nothing will ever call `wg.Done()`
* The `wg.Wait` is deadlocked

Reply:
This makes lots of sense, and it'd actually explain why I was still noticing some errors during tests that wouldn't properly clean up or reconcile everything before shutting stuff down (including the event recorder, which panics).